### PR TITLE
Fix false-positive in sigchanyzer check when os.Signal channel is unused

### PR DESCRIPTION
**Pull request by AGMETEOR**
Fixes [#45043](https://github.com/golang/go/issues/45043)

### CODE REVIEW

File: sigchanyzer.go [hunk 1]
```diff
@@ -48,17 +48,25 @@ func run(pass *analysis.Pass) (interface{}, error) {
 			return
 		}
 		var chanDecl *ast.CallExpr
+
+		// track whether call.Args[0] is a *ast.CallExpr. This is true when make(chan os.Signal) is passed directly to signal.Notify
+		wasCallExpr := false
```
Comment (odeke-em) [inline sigchanyzer.go:53]: Let's make this name more direct and more legible. Let's rename this to `inlinedChanMake := false`

File: sigchanyzer.go [hunk 2]
```diff
@@ -48,17 +48,25 @@ func run(pass *analysis.Pass) (interface{}, error) {
...
+
 		switch arg := call.Args[0].(type) {
 		case *ast.Ident:
 			if decl, ok := findDecl(arg).(*ast.CallExpr); ok {
 				chanDecl = decl
+				wasCallExpr = false
```
Comment (odeke-em) [inline sigchanyzer.go:59]: We don't need this.

File: sigchanyzer.go [hunk 3]
```diff
@@ -48,17 +48,25 @@ func run(pass *analysis.Pass) (interface{}, error) {
...
 			}
 		case *ast.CallExpr:
 			chanDecl = arg
+			wasCallExpr = true
 		}
-		if chanDecl == nil || len(chanDecl.Args) != 1 {
+
+		if chanDecl == nil || len(chanDecl.Args) != 1 || wasCallExpr {
 			return
 		}
```
Comment (odeke-em) [inline sigchanyzer.go:68]: What we could do is clearly separate these cases, so firstly
```go
                if chanDecl == nil || len(chanDecl.Args) != 1 {
                        return
                }
        
                // There is one exception. If we've got:
                // signal.Notify(make(chan os.Signal), signal)
                // The Go runtime will write to the channel with
                // a non-blocking send. See https://golang.org/issues/45043
                if nonBlockingChanSend {
                        return
                }
```

**Review by odeke-em — CHANGES_REQUESTED** [review]
Congrats @AGMETEOR, thank you for this change and welcome to Orijtech projects, we appreciate your contribution and look forward to you propagating this fix to the Go project.

Please update your commit message to be even more descriptive for example:
```shell
ignore inlined signal.Notify(make(os.Signal)) as the Go runtime delivers a non-blocking send
``` 

I've added some code changes, but otherwise, LGTM. Thank you and exciting!

File: sigchanyzer.go (hunk 2 above)
Comment (odeke-em) [reply sigchanyzer.go:59]: The reason why we don't need it is because the zero value was already
declared up above, and setting it to `false` again is unnecessary.

File: sigchanyzer.go [hunk 4]
```diff
@@ -48,17 +48,25 @@ func run(pass *analysis.Pass) (interface{}, error) {
...
+		// track whether call.Args[0] is a *ast.CallExpr. This is true when make(chan os.Signal) is passed directly to signal.Notify
+		wasCallExpr := false
+
 		switch arg := call.Args[0].(type) {
 		case *ast.Ident:
 			if decl, ok := findDecl(arg).(*ast.CallExpr); ok {
 				chanDecl = decl
+				wasCallExpr = false
 			}
 		case *ast.CallExpr:
```
Comment (cuonglm) [inline sigchanyzer.go:61]: We can just skip this case entirely, because `chanDecl` will be nil and the check below will be fired up. I think the main point left is whether we should skip this entirely, or just for `make` builtin?

File: sigchanyzer.go (hunk 3 above)
Comment (cuonglm) [reply sigchanyzer.go:68]: @odeke-em This is not exactly one exception, though. Above code will effectively ignore any call, not just the `make(chan os.Signal)`.

Also, the `nonBlockingChanSend` isn't meaningful. It's safe not because that runtime send to the channel using blocking send, but because the channel isn't read anywhere.

**Comment by odeke-em** [comment]
> Above code will effectively ignore any call, not just the make(chan
os.Signal).

Yup, we are relying on the type checker firstly to have succeeded in any
calls that get passed into signal.Notify. We only need to ensure that no
arguments are passed into that call, and that’s my approach.

**Comment by odeke-em** [comment]
Thanks for the suggestion and simplification, Cuong to remove the check for
all calls. We can go ahead and remove the switch for call expressions but
add a comment that  given that any calls passed into signal.Notify will
then be delegated to the type checker.

A problem though that I see with that approach is if a known non-buffered
channel is created above, but passed in through a closure, we would never
catch it.

ch := make(chan os.Signal)
f := func() chan os.Signal { return ch }
signal.Notify(f(), signal.Interrupt)

Or passed in through another function.
This case will very interesting, but we anyways need to  figure it out. To
solve it we’d need to do some more involved static assignment then
basically inline and detect the calls.

I think we should really just ensure that we are filtering “make” calls
only to distinctly handle the bug at hand, the rest can be dealt with on
their own.

Allan, please modify and ensure that in that switch we check exactly if
make(...) was invoked then proceed with the rest of the suggested changes.

**Comment by cuonglm** [comment]
@odeke-em That's the reason at first place, we conservatively treat any call to os.Notify as unsafe, because we don't know what the caller do with the channel. However, that case is rare, and shouldn't be passed in code review phase anyway.

**Comment by AGMETEOR** [comment]
Hey @odeke-em and @cuonglm , thanks for the review. Will implement this feedback tomorrow. Been a bit unwell

**Comment by AGMETEOR** [comment]
@odeke-em I have implemented the suggestions. Mind taking another look?

**Comment by AGMETEOR** [comment]
Hey @cuonglm , sorry I didn't quite understand what you suggested. Can you please explain further?

File: sigchanyzer.go [hunk 5]
```diff
@@ -48,17 +48,30 @@ func run(pass *analysis.Pass) (interface{}, error) {
...
 			chanDecl = arg
+			inlinedChanMake = true
 		}
+
 		if chanDecl == nil || len(chanDecl.Args) != 1 {
 			return
 		}
+
+		// There is one exception. If we've got: signal.Notify(make(chan os.Signal), signal)
+		// the Go runtime will write to the channel with a non-blocking send. See https://golang.org/issues/45043
```
Comment (cuonglm) [inline sigchanyzer.go:70]: This is not the main point. Whether you use `signal.Notify(make(chan os.Signal), signal)` or `signal.Notify(ch, signal)`, the runtime will always write to the channel using non-blocking send.

The former is safe, because nothing will read from the channel.

File: sigchanyzer.go [hunk 6]
```diff
@@ -48,17 +48,30 @@ func run(pass *analysis.Pass) (interface{}, error) {
 			return
 		}
 		var chanDecl *ast.CallExpr
+
+		// track whether call.Args[0] is a *ast.CallExpr. This is true when make(chan os.Signal) is passed directly to signal.Notify
```
Comment (cuonglm) [inline sigchanyzer.go:52]: This is true for all calls, not just `make(chan os.Signal)`.

**Review by cuonglm — CHANGES_REQUESTED** [review]

File: sigchanyzer.go (hunk 4 above)
Comment (AGMETEOR) [reply sigchanyzer.go:60]: @cuonglm didn't skip this and added a check to make sure the call is a make builtin i.e. ```make(chan os.Signal)``` or ```make(chan os.Signal, 1)```

File: sigchanyzer.go (hunk 6 above)
Comment (AGMETEOR) [reply sigchanyzer.go:52]: Took care of this

File: sigchanyzer.go [hunk 7]
```diff
@@ -132,3 +143,52 @@ func findDecl(arg *ast.Ident) ast.Node {
 	}
 	return nil
 }
+
+func isMakeCallExpr(info *types.Info, call *ast.CallExpr) bool {
```
Comment (cuonglm) [inline sigchanyzer.go:147]: This seems too complicated, we can do something like:

```
func isBuiltinMake(info *types.Info, call *ast.CallExpr) bool {
	typVal := info.Types[call.Fun]
	if !typVal.IsBuiltin() {
		return false
	}
	switch fun := call.Fun.(type) {
	case *ast.Ident:
		return info.ObjectOf(fun).Name() == "make"
	default:
		return false
	}
}		
```

File: sigchanyzer.go (hunk 6 above)
Comment (cuonglm) [reply sigchanyzer.go:52]: I think `inlinedChanMake` is confused, just remove this, and do something like:

```
		var chanDecl *ast.CallExpr
		switch arg := call.Args[0].(type) {
		case *ast.Ident:
			if decl, ok := findDecl(arg).(*ast.CallExpr); ok {
				chanDecl = decl
			}
		case *ast.CallExpr:
			// Only signal.Notify(make(chan os.Signal), os.Interrupt) is safe,
			// conservatively treate others as not safe, see golang/go#45043
			if isBuiltinMake(pass.TypesInfo, arg) {
				return
			}
			chanDecl = arg
		}
```

**Review by cuonglm — CHANGES_REQUESTED** [review]

**Comment by AGMETEOR** [comment]
@cuonglm thanks a lot for the review. I have changed the code to use your suggestions.

File: sigchanyzer.go [hunk 8]
```diff
@@ -48,17 +48,25 @@ func run(pass *analysis.Pass) (interface{}, error) {
 			return
 		}
 		var chanDecl *ast.CallExpr
+
```
Comment (cuonglm) [inline sigchanyzer.go:51]: Nit: remove this blank line.

**Review by cuonglm — APPROVED** [review]

File: sigchanyzer.go (hunk 8 above)
Comment (AGMETEOR) [reply sigchanyzer.go]: Done.

**Comment by odeke-em** [comment]
Well done @AGMETEOR. Thank you for the review @cuonglm. Next steps @AGMETEOR, please send your first CL (🎉👏🏿💯) to golang.org/x/tools with these exact changes and tag us, we’ll fast track it and once that’s merged in, then vendor into Go and you’d have completed your entire pipeline of Go changes. Congratulations on this very first step.

**Review by odeke-em — APPROVED** [review]
